File: 3398.smallest-substring-with-identical-characters-i.java
#
# @lc app=leetcode id=3398 lang=java
#
# [3398] Smallest Substring With Identical Characters I
#

# @lc code=start
class Solution {
    public int minLength(String s, int numOps) {
        int n = s.length();
        int left = 1, right = n;
        while (left < right) {
            int mid = (left + right) / 2;
            if (canAchieve(s, numOps, mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
    private boolean canAchieve(String s, int numOps, int maxLen) {
        int flips = 0;
        int n = s.length();
        int i = 0;
        while (i < n) {
            int j = i;
            while (j < n && s.charAt(j) == s.charAt(i)) j++;
            int len = j - i;
            if (len > maxLen) {
                // Need to break this run into pieces of at most maxLen
                flips += (len - 1) / maxLen;
            }
            i = j;
        }
        return flips <= numOps;
    }
}
# @lc code=end

File: 3398.smallest-substring-with-identical-characters-i.go
#
# @lc app=leetcode id=3398 lang=golang
#
# [3398] Smallest Substring With Identical Characters I
#

# @lc code=start
func minLength(s string, numOps int) int {
    n := len(s)
    maxLen := 0 // This will track the maximum length of a substring with identical characters
    left := 0   // Left pointer for our sliding window
    count0, count1 := 0, 0 // Count of '0's and '1's in current window
    
    for right := 0; right < n; right++ {
        if s[right] == '0' {
            count0++
        } else {
            count1++
        }
        
        // If we need more than numOps flips to make all chars identical in this window
        while min(count0, count1) > numOps {
            if s[left] == '0' {
                count0--
            } else {
                count1--
            }
            left++
        }
        
        // Calculate current maxLen as longest substring of identical chars within current valid window
        maxLen = max(maxLen, right - left + 1)
    }
    
    return n - maxLen // Minimum length after minimizing longest identical character substring
}

func min(a, b int) int { if a < b { return a }; return b }
func max(a, b int) int { if a > b { return a }; return b }
# @lc code=end